crates/accelerate/src/circuit_library/quantum_volume.rs

```diff
@@ -0,0 +1,149 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use std::thread::available_parallelism;
+
+use qiskit_circuit::imports::UNITARY_GATE;
+
+use pyo3::prelude::*;
+
+use crate::getenv_use_multiple_threads;
+use faer_ext::{IntoFaerComplex, IntoNdarrayComplex};
+use ndarray::prelude::*;
+use num_complex::Complex64;
+use numpy::IntoPyArray;
+use rand::prelude::*;
+use rand_distr::Normal;
+use rand_pcg::Pcg64Mcg;
+use rayon::prelude::*;
+
+use qiskit_circuit::circuit_data::CircuitData;
+use qiskit_circuit::operations::Param;
+use qiskit_circuit::operations::PyInstruction;
+use qiskit_circuit::packed_instruction::PackedOperation;
+use qiskit_circuit::Qubit;
+use smallvec::smallvec;
+
+#[inline]
+fn random_unitaries(seed: u64, size: usize) -> impl Iterator<Item = Array2<Complex64>> {
+    let mut rng = Pcg64Mcg::seed_from_u64(seed);
+    let dist = Normal::new(0., 1.0).unwrap();
+
+    (0..size).map(move |_| {
+        let mut z: Array2<Complex64> = Array2::from_shape_simple_fn((4, 4), || {
+            Complex64::new(dist.sample(&mut rng), dist.sample(&mut rng))
+        });
+        z.mapv_inplace(|x| x * std::f64::consts::FRAC_1_SQRT_2);
+        let qr = z.view().into_faer_complex().qr();
+        let r = qr.compute_r().as_ref().into_ndarray_complex().to_owned();
+        let mut d = r.into_diag();
+        d.mapv_inplace(|d| d / d.norm());
```

Comment:
This does an in-place substitution on the diagonal `d`  so that each value also named `d` (which is probably why this is confusing) is replaced with `d / d.norm()` for each element `d`. I'll rename the diagonal to `diag` or something to make it clear which is an element and which is the array.

Reply:
I am wondering if I should rewrite this using fixed size arrays or nalgebra's `Matrix2<Complex64>` to make this all faster. Right now I'm pretty confident the overhead is mostly dealing with allocations for 2x2 arrays.

Reply:
I did this here: https://github.com/Qiskit/qiskit/pull/13238/commits/32b4c1f8c32a7e531c9bc4b12288be89b3cc5b38 the code is a lot more verbose now, but I think it's a bit more explicit and it'll be a bit more optimized (although the random unitary construction is not the bottleneck).

Reply:
Nice! The bottleneck is the `UnitaryGate` creation Python-side as you wrote above, right? I think a 10x speedup (maybe more now with the explicit matrix allocation?) is already great, plus we have the circuit as function reaching the goal of #13046 😄 But if you'd like to test with pre-generating the gates I'm also happy to wait 🙂

Reply:
Yeah, we are totally bottlenecked by the `UnitaryGate` object creation, I ran a profile with the current state of the PR running the test script I was using to generate the data for the graph I shared above. In serial mode (py-spy doesn't handle the multithreading well) and the `UnitaryGate` object creation is taking ~58% of the time, while the random unitary generation is only ~30% of the runtime (almost all of that is performing the qr factorization).

I'll do a quick test of serial mode dumping the unitaries to a vec first to see if that makes a difference performance wise (I don't expect it to), and will leave a comment here with the results.